sigstore/_internal/sct.py

```diff
@@ -16,75 +16,124 @@
 """
 
+import hashlib
+import logging
 import struct
+from typing import List, Union
 
-import cryptography.hazmat.primitives.asymmetric.ec as ec
+import cryptography.hazmat.primitives.asymmetric.padding as padding
 from cryptography.exceptions import InvalidSignature
 from cryptography.hazmat.primitives import hashes, serialization
-from cryptography.x509 import Certificate
+from cryptography.hazmat.primitives.asymmetric import ec, rsa
+from cryptography.x509 import Certificate, ExtendedKeyUsage, ObjectIdentifier
+from cryptography.x509.certificate_transparency import (
+    SignedCertificateTimestamp,
+)
 
-from sigstore._internal.fulcio import DetachedFulcioSCT
+logger = logging.getLogger(__name__)
 
 
-def _pack_digitally_signed(sct: DetachedFulcioSCT, cert: Certificate) -> bytes:
+def _pack_digitally_signed(
+    sct: SignedCertificateTimestamp,
+    cert: Certificate,
+    issuer_key_hash: bytes,
+) -> bytes:
     """
     The format of the digitally signed data is described in IETF's RFC 6962.
-
-    1 SCT Version
-    1 Signature Type
-    8 Timestamp
-    2 Entry Type
-    3 Certificate Length
-    X Certificate Data
-    2 Extensions Length
     """
 
     # The digitally signed format requires the certificate in DER format.
-    cert_der: bytes = cert.public_bytes(encoding=serialization.Encoding.DER)
+    cert_der: bytes = cert.tbs_precertificate_bytes
 
-    # The length should then be split into three bytes.
+    # The length is a u24, which isn't directly supported. So we have to
+    # decompose it into 3 bytes.
     unused, len1, len2, len3 = struct.unpack(
         "!4B",
         struct.pack("!I", len(cert_der)),
     )
     if unused:
         raise InvalidSctError(f"Unexpectedly large certificate length: {len(cert_der)}")
 
+    # No extensions are currently specified.
+    if len(sct.extension_bytes) != 0:
+        raise InvalidSctError("Unexpected trailing extension bytes")
+
+    # NOTE(ww): This is incorrect for non-embedded SCTs, since it unconditionally
+    # embeds the issuer_key_hash.
+
     # Assemble a format string with the certificate length baked in and then pack the digitally
     # signed data
-    pattern = "!BBQhBBB%ssh" % len(cert_der)
+    # fmt: off
+    pattern = "!BBQH32sBBB%ssH" % len(cert_der)
     data = struct.pack(
         pattern,
-        sct.version.value,
-        0,  # Signature Type
-        int(sct.timestamp.timestamp() * 1000),
-        0,  # Entry Type
-        len1,
-        len2,
-        len3,
-        cert_der,
-        len(sct.extensions),
+        sct.version.value,                      # sct_version
+        0,                                      # signature_type (certificate_timestamp(0))
+        int(sct.timestamp.timestamp() * 1000),  # timestamp (milliseconds)
+        sct.entry_type.value,                   # entry_type (x509_entry(0) | precert_entry(1))
+        issuer_key_hash,                        # issuer_key_hash[32]
+        len1,                                   # \
```

Comment:
(Feel free to ignore if you've already validated the contents here) What are each of these three lengths?

Reply:
Yep, I have a unit test with testvectors validating this. They're the individual bytes of the `uint24_t` for the following cert/TBS data, since `struct` unfortunately doesn't support packing/unpacking 24-bit integers directly.